torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -0,0 +1,442 @@
+  if (input.size() > numGPUs_) {
+    throw std::runtime_error(
+        "The number of input tensors is larger than "
+        "the number of available GPUs");
+  }
+
+  // To make sure each tensor is on separate devices
+  std::unordered_set<int> usedDevices;
+  usedDevices.reserve(input.size());
+
+  uint64_t inputNumElement = input[0].numel();
+  auto elementType = input[0].type().scalarType();
+
+  for (auto i = 0; i < input.size(); ++i) {
+    //  Check to make sure it's a GPU dense tensor
+    if (!(input[i].type().is_cuda() && !input[i].type().is_sparse() &&
+          output[i].type().is_cuda() && !output[i].type().is_sparse())) {
+      throw std::runtime_error(
+          "Only CUDA dense tensor is supported for NCCL "
+          "collective operations");
+    }
+    // Check the tensor type is identical
+    if (input[i].type().scalarType() != elementType ||
+        output[i].type().scalarType() != elementType) {
+      throw std::runtime_error(
+          "Expecting all GPU tensors to have identical "
+          "type");
+    }
+    // Check the input tensor size is identical
+    if (input[i].numel() != inputNumElement) {
+      throw std::runtime_error(
+          "Expecting all input tensors to have identical "
+          "number of elements");
+    }
+    // Check the output tensor size equals to input tensor size
+    if (output[i].numel() != inputNumElement * outputOverInput) {
+      throw std::runtime_error(
+          "The number of elements of output tensor does "
+          "not match the number of elements of the input "
+          "tensor");
+    }
+    // Contiguous verification
+    if (!input[i].is_contiguous() || !output[i].is_contiguous()) {
+      throw std::runtime_error("Expecting all GPU tensors to be contiguous");
+    }
+
+    bool inserted;
+    std::tie(std::ignore, inserted) = usedDevices.insert(input[i].get_device());
+    // Device verification, if the insertion didn't take place
+    if (!inserted) {
+      throw std::runtime_error("Expecting inputs on different GPU devices");
+    }
+
+    // Now check the output device
+    if (input[i].get_device() != output[i].get_device()) {
+      throw std::runtime_error(
+          "Expecting input and output tensors to be on "
+          "the same device");
+    }
+  }
+}
+
+std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+  tensorCheckHelper(tensors, tensors);
+
+  auto devices = getDevicesOfTensors(tensors);
+  auto key = getKeyFromDevices(devices);
+  // Get the NCCL communicators
+  auto& ncclComms = getNCCLComm(key, devices);
+
+  // First let NCCL streams wait for THC stream
+  syncStreams(thcState_, devices, ncclEvents_[key], ncclStreams_[key]);
+
+  // Work itself will create the CUDA events on all GPUs of tensors
+  auto work = std::make_shared<ProcessGroupNCCL::WorkNCCL>(devices);
+
+  // Guard GPU device
+  CUDADevice gpuGuard;
+
+  // Queue the NCCL kernel
```

Comment:
Comment can go and I think also is not correct; my understanding is that the kernels are only queued when you run ncclGroupEnd()